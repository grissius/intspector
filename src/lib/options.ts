import * as fs from 'fs';
import * as ts from 'typescript';

/** @ignore */
let options: ts.CompilerOptions;

/**
 * @see https://github.com/Microsoft/TypeScript/issues/5276#issuecomment-148926002
 * @ignore
 */
function convertConfigToCompilerOptions(opts: {
    compilerOptions: ts.CompilerOptions;
}) {
    const { options, errors } = ts.parseJsonConfigFileContent(
        {
            ...opts,
            // if files are not specified then parseJsonConfigFileContent
            // will use ParseConfigHost to collect files in containing folder
            files: [],
        },
        ts.sys,
        ''
    );
    // Remove the following error:
    // { messageText: "The 'files' list in config file 'tsconfig.json' is empty.", category: 1, code: 18002 }
    const relevantErrors = errors.filter(e => e.code !== 18002);
    if (relevantErrors.length > 0) {
        throw relevantErrors;
    }

    return options;
}

/**
 * Set compiler options for inspect functions.
 * Operates on global object.
 *
 * @param input Compiler options. Beware, enum option values are different than the ones from your json.
 * @param ignoreProjectOptions If true, do not merge the passed options with the project options (from tsconfig.json) but only use the passed options exclusively. False by default.
 * @returns current options
 */
export const setOptions = (
    input: ts.CompilerOptions,
    ignoreProjectOptions = false
) => {
    if (ignoreProjectOptions) {
        options = input;
        return options;
    }
    const maybeFile = ts.findConfigFile(__dirname, fs.existsSync);
    if (maybeFile === undefined) {
        throw new Error('setOptions: Cannot find tsconfig.json');
    }
    const { config, error } = ts.readConfigFile(maybeFile, path =>
        fs.readFileSync(path).toString()
    );
    const parsedConfig = convertConfigToCompilerOptions(config);
    if (error !== undefined) {
        const message = `TS${error.code}: ${error.file}:${error.start} ${error.messageText}`;
        throw new Error(message);
    }
    options = { ...parsedConfig, ...input };
    return options;
};

/**
 * Get the TypeScript compiler options to be used for inspect functions.
 */
export const getOptions = (): ts.CompilerOptions => {
    if (options === undefined) {
        setOptions({});
    }
    return options;
};
