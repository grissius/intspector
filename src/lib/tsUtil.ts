import * as fs from 'fs';
import * as ts from 'typescript';

export const FILENAME = '__ts-type-test-inline-e1d70ff1__';
export const FILENAME_RE = new RegExp(FILENAME);
export const TYPENAME_RE = /type\s+(__[a-zA-Z0-9][_a-zA-Z0-9]*)[^_a-zA-Z0-9]+/g;

/**
 * The global options object which setOptions and getOptions work on
 */
let options: ts.CompilerOptions;

/**
 * Set compiler options for checking inline code
 *
 * These options set a global options object which any subsequent
 * invocations of `resolveTypes` will use.
 *
 * ignoreProjectOptions controls if the options from tsconfig are
 * used or not.
 *
 * @param input Compiler options for use when checking inline code.
 * @param ignoreProjectOptions If true, do not merge the passed options with
 *     the project options (from tsconfig.json) but only use the passed options
 *     exclusively. False by default.
 */
export const setOptions = (
    input: ts.CompilerOptions,
    ignoreProjectOptions = false
): void => {
    if (ignoreProjectOptions) {
        options = input;
        return;
    }
    const maybeFile = ts.findConfigFile(__dirname, fs.existsSync);
    if (maybeFile === undefined) {
        throw new Error('setOptions: Cannot find tsconfig.json');
    }
    const { config, error } = ts.readConfigFile(maybeFile, path =>
        fs.readFileSync(path).toString()
    );
    if (error !== undefined) {
        const message = `TS${error.code}: ${error.file}:${error.start} ${
            error.messageText
        }`;
        throw new Error(message);
    }
    options = { ...config, ...input };
};

/**
 * Get the TypeScript compiler options to be used for resolving types
 */
export const getOptions = (): ts.CompilerOptions => {
    if (options === undefined) {
        setOptions({});
    }
    return options;
};

const createInlineProgram = (code: string) => {
    // Work around definite assignemt checking: inlineSourceFile is assigned
    // when ts.createProgram is created
    let inlineSourceFile!: ts.SourceFile;
    const getSourceFile = (
        fileName: string,
        languageVersion: ts.ScriptTarget,
        ...args: any[]
    ) => {
        if (!FILENAME_RE.test(fileName)) {
            return (compilerHost.getSourceFile as any)(
                fileName,
                languageVersion,
                ...args
            );
        }
        if (inlineSourceFile === undefined) {
            inlineSourceFile = ts.createSourceFile(
                FILENAME,
                code,
                languageVersion
            );
        }
        return inlineSourceFile;
    };
    const options = getOptions();
    const compilerHost = ts.createCompilerHost(options);
    const customCompilerHost = {
        ...compilerHost,
        getSourceFile,
    };
    const program = ts.createProgram([FILENAME], options, customCompilerHost);
    return { program, inlineSourceFile };
};

export const resolveTypes = (
    code: string,
    whitelist: string[],
    // types?: Record<string, string>
) => {
    const { program, inlineSourceFile } = createInlineProgram(code);
    const checker = program.getTypeChecker();

    checker
        .getSymbolsInScope(
            inlineSourceFile!.endOfFileToken,
            ts.SymbolFlags.Type
        )
        .filter(symbol => whitelist.includes(symbol.name))
        .forEach(symbol => {
            let typeAsString: string; // cache value once computed
            const type = checker.getDeclaredTypeOfSymbol(symbol);
            typeAsString = checker.typeToString(
                type,
                inlineSourceFile,
                ts.TypeFormatFlags.InTypeAlias
            );
            console.log(typeAsString);
            return typeAsString;
        });

    let diagnostics: ReadonlyArray<ts.Diagnostic>;

    return {
        types: {},
        get diagnostics() {
            if (diagnostics === undefined) {
                diagnostics = [
                    ...program.getSyntacticDiagnostics(inlineSourceFile),
                    ...program.getSemanticDiagnostics(inlineSourceFile),
                    ...program.getDeclarationDiagnostics(inlineSourceFile),
                ];
            }
            return diagnostics;
        },
    };
};

const p = resolveTypes("type x = Pick<{a: string, b: number}, 'b'>", ['x']);
console.log(p);
