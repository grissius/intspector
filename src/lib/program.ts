import * as ts from 'typescript';
import { getOptions } from './options';

export const FILENAME = '__ts-type-test-inline-e1d70ff1__';
export const FILENAME_RE = new RegExp(FILENAME);
export const TYPENAME_RE = /type\s+(__[a-zA-Z0-9][_a-zA-Z0-9]*)[^_a-zA-Z0-9]+/g;

export const createInlineProgram = (code: string) => {
    // Work around definite assignemt checking: inlineSourceFile is assigned
    // when ts.createProgram is created
    let inlineSourceFile!: ts.SourceFile;
    const getSourceFile = (
        fileName: string,
        languageVersion: ts.ScriptTarget,
        ...args: any[]
    ) => {
        if (!FILENAME_RE.test(fileName)) {
            return (compilerHost.getSourceFile as any)(
                fileName,
                languageVersion,
                ...args
            );
        }
        if (inlineSourceFile === undefined) {
            inlineSourceFile = ts.createSourceFile(
                FILENAME,
                code,
                languageVersion
            );
        }
        return inlineSourceFile;
    };
    const options = getOptions();
    const compilerHost = ts.createCompilerHost(options);
    const customCompilerHost = {
        ...compilerHost,
        getSourceFile,
    };
    const program = ts.createProgram([FILENAME], options, customCompilerHost);
    return { program, inlineSourceFile };
};