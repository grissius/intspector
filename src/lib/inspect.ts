import * as ts from 'typescript';
import { createInlineProgram } from './program';

export const inspect = <T extends Record<string, string>>(
    preamble: string,
    typeMap: T
) => {
    const code = [preamble].concat(Object.entries(typeMap).map(([k, d]) => `type ${k} = ${d}`)).join(' ')
    const { program, inlineSourceFile } = createInlineProgram(code);
    const checker = program.getTypeChecker();

    const pairs = checker
        .getSymbolsInScope(
            inlineSourceFile!.endOfFileToken,
            ts.SymbolFlags.Type
        )
        .filter(symbol => Object.keys(typeMap).includes(symbol.name))
        .map(symbol => {
            const symbolType = checker.getDeclaredTypeOfSymbol(symbol);
            const type = checker.typeToString(
                symbolType,
                inlineSourceFile,
                ts.TypeFormatFlags.InTypeAlias
            );
            return [symbol.name, type];
        });

    const types = pairs.reduce((acc, [k, v]) => {
        acc[k] = v;
        return acc;
    }, {} as { [K in keyof T]: string })

    const diagnostics = [
        ...program.getSyntacticDiagnostics(inlineSourceFile),
        ...program.getSemanticDiagnostics(inlineSourceFile),
        ...program.getDeclarationDiagnostics(inlineSourceFile),
    ];

    if (diagnostics.length > 0) {
        throw diagnostics;
    }

    return types;
};

const p = inspect('', { x: `Pick<{a: string, b: number}, 'b'>`});
console.log(p);
