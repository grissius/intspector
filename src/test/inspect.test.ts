import { inspectObjectCore } from 'lib/inspect';

describe('inspectObjectCore', () => {
    test('Basic types', () => {
        const types = inspectObjectCore('', {
            str: 'string',
            anys: 'any[]',
            record: 'Record<number, string>',
            fn: '(x: boolean) => number',
            id: '<T>(x: T) => T',
        });
        expect(types).toMatchSnapshot();
    });
    test('Self-reference', () => {
        const types = inspectObjectCore('', {
            a: 'string',
            b: 'a | Promise<a>',
        });
        expect(types).toMatchSnapshot();
    });
    test('Preamble', () => {
        const types = inspectObjectCore("let x = 1; let y = '';", {
            res: 'typeof x & typeof y',
        });
        expect(types).toMatchSnapshot();
    });
    test('Import', () => {
        const types = inspectObjectCore(
            "import { inspectObject } from 'index';",
            {
                res: 'typeof inspectObject',
            }
        );
        expect(types).toMatchSnapshot();
    });
    test('Bad import', () => {
        expect(() =>
            inspectObjectCore("import { foo } from 'blah';", {
                res: 'typeof foo',
            })
        ).toThrow();
    });
});
