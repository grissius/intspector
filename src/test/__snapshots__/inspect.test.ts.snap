// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspect inspectObjectCore Basic types 1`] = `
Object {
  "anys": "any[]",
  "fn": "(x: boolean) => number",
  "id": "<T>(x: T) => T",
  "record": "{ [x: number]: string; }",
  "str": "string",
}
`;

exports[`inspect inspectObjectCore Import 1`] = `
Object {
  "res": "<T extends string | Record<string, string>>(input: T) => T",
}
`;

exports[`inspect inspectObjectCore Preamble 1`] = `
Object {
  "res": "number & string",
}
`;

exports[`inspect inspectObjectCore Self-reference 1`] = `
Object {
  "a": "string",
  "b": "string | Promise<string>",
}
`;
